### ISSUE TITLE
Performance issue with <cms:thumbnail>

### ISSUE THREAD

**Issue by mcarneiro** [CouchCMS/CouchCMS#15]
Hi guys, I have a situation where I have lots of thumbnails being generated on a page.

I have a list of images with varations for responsive layout. So in my example I have 17 images and a logic that generate thumbs for 3 different sizes + 2x (retina display) version for each image and output a CSS media queries inside `<head>`. So I have 102 different images.

I was experiencing slow page load, between 10 and 12 seconds. While debugging, I saw that inside `couch/includes/timthumb.php`, the `if` that verifies the thumb exists is after the moment where image is opened.

* https://github.com/CouchCMS/CouchCMS/blob/master/couch/includes/timthumb.php#L59 - opening the file;
* https://github.com/CouchCMS/CouchCMS/blob/master/couch/includes/timthumb.php#L126 - return if thumb exists;

I got the content of the `if` in line 126 and placed it right before line 59. Everything is working fine in my project and much faster, as the image doesn't need to be opened when the thumbnail is already generated.

The page load time here dropped from ≈11s to ≈0.5s. Maybe that's something to implement in the source code? (I can create a PR for this) I couldn't see any possible problem with this solution as before it just check some size and make some calculations.

**Comment by mcarneiro** [comment]
Forked and created a branch with it: https://github.com/CouchCMS/CouchCMS/compare/master...mcarneiro:fix-performance-timthumb?expand=1

If everything is ok, i'll send the PR.

Cheers

**Comment by kksidd** [comment]
Thanks :)

I'll take that your code is using <cms:thumbnail /> tag, right?

The **$new_width** and **$new_height** parameter we use to craft the thumbnail name can vary - specifically when either of the two params is provided as 0 or the **$enforce_max** is specified as '1'. In all such cases the source image has to be opened to work out the right values for the new_height and new_width parameter.

As such, we cannot make the amendment you proposed to affect all thumbnails.

I think we should apply this optimization only for the thumnails that don't need the source image.
The check could be like -
```
if( $new_width && $new_height && !$enforce_max ){
    // check if the thumbnail already exists and go no further if it does..
}
```
I also think that, instead of within timthumb.php, it'd be more appropriate if we placed this check within the <cms:thumbnail /> tag (in tags.php) instead.

Your thoughts? Also, could you please try making the suggested change to cms:thumbnail?

Thanks.

**Comment by mcarneiro** [comment]
Mmmm, good point. I'll take a look and think of another solution and post it here on monday :)

**Comment by mcarneiro** [comment]
Hey guys, today I played around with this issue and made better tests to check the performance and implications of the code change.

While I was checking, I realized that the performance impact was bigger when the image was bigger, so I realized that maybe `open_image` wasn't the best way to just get the dimensions of the image.

I'm not a PHP developer, so I googled about faster ways to get image dimensions and found the native `getimagezise` function (for local files). I made some tests here and it's much faster now. Now I just define `$image` when it needs to resize / crop / write the new version.

So in my test case I have 17 images and I run the following calls:

```
<cms:thumbnail width='100' height='100' /> 
<cms:thumbnail width='150' /> 
<cms:thumbnail height='90' /> 
<cms:thumbnail width='200' height='100' enforce_max='1' /> 
<cms:thumbnail width='150' height='200' enforce_max='1' /> 
<cms:thumbnail width='300' enforce_max='1' /> 
<cms:thumbnail height='230' enforce_max='1' /> 
```

My results where (file format / sizes):

**JPG 3840 x 899px**

* Original: ≈22s (first page-load), ≈12s (second page-load);
* Using `getimagesize`: ≈18s (first), ≈100ms (second);

**PNG 1920 x 1080px**

* Original: ≈23s (first), ≈14s (second);
* Using `getimagesize`: ≈21s (first), ≈100ms (second);

**GIF 256 x 200px**

* Original: ≈2.7s (first), ≈450ms (second);
* Using `getimagesize`: ≈2s (first), ≈100ms (second);

Basically you can feel the real difference when using large images (which is more common today because of the retina display. eg.: very common to have a full-screen image for iPad with 2048 x 1536px) and after the first load it's always consistent around 100ms.

Also, I tested with both JPG and PNG optimized using tinypng.com, to strip all exif and metadata information (just to check if `getimagesize` relied someway on these infos).

Do you know if is there any limitation on using `getimagesize` method instead of `imagecreatefromXXX` from `open_image`?

Also, I changed inside timthumb.php because I didn't want to duplicate the size verification inside `<cms:thumbnail>` tag. Take a look at the change I made and tell me what you think: https://github.com/mcarneiro/CouchCMS/commit/8ea928db6f16fa902a526e593762261b1787d6a6

**Comment by kksidd** [comment]
Thanks for posting your observations. Very useful.

There are no limitations in using getimagesize, that I know of.
I'll do some more tests of my own and let you know.

**Comment by cheesypoof** [comment]
Can this be closed now that 074a8a8 has made it in to the repository?
